lib/assert.js

```diff
@@ -608,6 +612,139 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new Set(),
+) {
+  function isPlainObject(obj) {
+    return typeof obj === 'object' && obj.constructor === Object;
+  }
+
+  // Check non object types equality
+  if (!isPlainObject(actual) || !isPlainObject(expected)) {
+    if (isDeepEqual === undefined) lazyLoadComparison();
+    return loose ? isDeepEqual(actual, expected) : isDeepStrictEqual(actual, expected);
+  }
+
+  // Check if actual and expected are null or not objects
+  if (actual == null || expected == null) {
+    return false;
+  }
+
+  // Check for Map object equality
+  if (actual instanceof Map && expected instanceof Map) {
+    if (actual.size !== expected.size) return false;
+    for (const { 0: key, 1: val } of actual) {
+      if (!expected.has(key)) return false;
+      if (!compareBranch(val, expected.get(key), loose, comparedObjects))
+        return false;
+    }
+    return true;
+  }
+
+  // Check for Set object equality
+  if (actual instanceof Set && expected instanceof Set) {
+    if (actual.size !== expected.size) return false;
+    for (const item of actual) {
+      if (!expected.has(item)) return false;
```

Comment:
Same here for `Set`s
```suggestion
  if (isSet(actual) && isSet(expected)) {
    if (actual.size !== expected.size) return false;
    const safeIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual);
    for (const item of safeIterator) {
      if (!SetPrototypeHas(expected, item)) return false;
```